File: javascript/src/TreeRegexp.ts
import execWithIndices from 'regexp-match-indices'

import Group from './Group.js'
import GroupBuilder from './GroupBuilder.js'

export default class TreeRegexp {
  public readonly regexp: RegExp
  public readonly groupBuilder: GroupBuilder

  constructor(regexp: RegExp | string) {
    if (regexp instanceof RegExp) {
      this.regexp = regexp
    } else {
      this.regexp = new RegExp(regexp)
    }
    this.groupBuilder = TreeRegexp.createGroupBuilder(this.regexp)
  }

  private static createGroupBuilder(regexp: RegExp): GroupBuilder {
    const source = regexp.source
    const stack: GroupBuilder[] = [new GroupBuilder()]
    const groupStartStack: number[] = []
    let escaping = false
    let charClass = false

    for (let i = 0; i < source.length; i++) {
      const c = source[i]
      if (c === '[' && !escaping) {
        charClass = true
      } else if (c === ']' && !escaping) {
        charClass = false
      } else if (c === '(' && !escaping && !charClass) {
        groupStartStack.push(i)
        const nonCapturing = TreeRegexp.isNonCapturing(source, i)
        const groupBuilder = new GroupBuilder()
        if (nonCapturing) {
          groupBuilder.setNonCapturing()
        }
        stack.push(groupBuilder)
      } else if (c === ')' && !escaping && !charClass) {
        const gb = stack.pop()
        if (!gb) throw new Error('Empty stack')
        const groupStart = groupStartStack.pop()
        if (gb.capturing) {
          gb.source = source.substring((groupStart || 0) + 1, i)
          stack[stack.length - 1].add(gb)
        } else {
          gb.moveChildrenTo(stack[stack.length - 1])
        }
      }
      escaping = c === '\\' && !escaping
    }
    const result = stack.pop()
    if (!result) throw new Error('Empty stack')
    return result
  }

  private static isNonCapturing(source: string, i: number): boolean {
    // Regex is valid. Bounds check not required.
    if (source[i + 1] !== '?') {
      // (X)
      return false
    }
    if (source[i + 2] !== '<') {
      // (?:X)
      // (?=X)
      // (?!X)
      return true
    }
    // (?<=X) or (?<!X) else (?<name>X)
    return source[i + 3] === '=' || source[i + 3] === '!'
  }

  public match(s: string): Group | null {
    const match = execWithIndices(this.regexp, s)
    if (!match) {
      return null
    }
    let groupIndex = 0
    const nextGroupIndex = () => groupIndex++
    return this.groupBuilder.build(match, nextGroupIndex)
  }
}

File: dart/lib/src/tree_regexp.dart
import 'package:cucumber_expressions/src/group.dart';
import 'package:cucumber_expressions/src/group_builder.dart';

/// Wraps a [RegExp] and builds a tree of capture [Group]s that mirrors the
/// nesting of the regular expression's groups.
class TreeRegexp {
  /// Creates a tree regexp from an existing [regexp].
  TreeRegexp(this.regexp) : groupBuilder = _createGroupBuilder(regexp);

  /// Creates a tree regexp by compiling [pattern].
  TreeRegexp.fromString(String pattern) : this(RegExp(pattern));

  /// The underlying regular expression.
  final RegExp regexp;

  /// The root group builder describing the capture group structure.
  final GroupBuilder groupBuilder;

  static GroupBuilder _createGroupBuilder(RegExp regexp) {
    final source = regexp.pattern;
    final stack = <GroupBuilder>[GroupBuilder()];
    final groupStartStack = <int>[];
    var escaping = false;
    var charClass = false;

    for (var i = 0; i < source.length; i++) {
      final c = source[i];
      if (c == '[' && !escaping) {
        charClass = true;
      } else if (c == ']' && !escaping) {
        charClass = false;
      } else if (c == '(' && !escaping && !charClass) {
        groupStartStack.add(i);
        final nonCapturing = _isNonCapturing(source, i);
        final groupBuilder = GroupBuilder();
        if (nonCapturing) {
          groupBuilder.setNonCapturing();
        }
        stack.add(groupBuilder);
      } else if (c == ')' && !escaping && !charClass) {
        final gb = stack.removeLast();
        final groupStart = groupStartStack.removeLast();
        if (gb.capturing) {
          gb.source = source.substring(groupStart + 1, i);
          stack.last.add(gb);
        } else {
          gb.moveChildrenTo(stack.last);
        }
      }
      escaping = c == r'\' && !escaping;
    }
    return stack.removeLast();
  }

  static bool _isNonCapturing(String source, int i) {
    // Regex is valid. Bounds check not required.
    if (i + 1 >= source.length || source[i + 1] != '?') {
      // (X)
      return false;
    }
    if (i + 2 >= source.length || source[i + 2] != '<') {
      // (?:X)
      // (?=X)
      // (?!X)
      return true;
    }
    // (?<=X) or (?<!X) else (?<name>X)
    return source[i + 3] == '=' || source[i + 3] == '!';
  }

  /// Matches [s] and returns the root [Group], or `null` if there is no match.
  Group? match(String s) {
    final match = regexp.firstMatch(s);
    if (match == null) {
      return null;
    }
    var groupIndex = 0;
    int nextGroupIndex() => groupIndex++;
    return groupBuilder.build(match, nextGroupIndex);
  }
}
